docs/developer_guide/style_guidelines_and_tests.md

```diff
@@ -1,35 +1,94 @@
+Type hints should be used wherever possible, and should use the most specific type accepted by the method.
+
+Example:
+
+```python
+def load_processed_dataset(
+        self,
+        split
+) -> Union[pd.DataFrame, Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]]:
+    """Loads the processed Parquet into a dataframe.
+
+    :param split: Splits along 'split' column if present.
+    :returns: The preprocessed dataset, or a tuple of (train, validation, test).
+    """
+```
+
+Functions with no arguments or return value may have a single-line docstring, ie:
+
+```python
+@pytest.fixture()
+def csv_filename():
+    """Yields a csv filename for holding temporary data."""
+```
 
 # Tests
 
-We are using `pytest` to run tests.
-To install all the required dependencies for testing, please do `pip install ludwig[test]`.
-Current test coverage is limited to several integration tests which ensure end-to-end functionality but we are planning to expand it.
+Ludwig uses two types of tests: unit tests and integration tests. Unit tests test a single module, and should
+individually be very fast. Integration tests run an end-to-end test of a single Ludwig functionality, like hyperopt or
+visualization. Ludwig tests are organized in the following directories:
+
+```
+├── ludwig                 - Ludwig library source code
+└── tests
+    ├── integration_tests  - End-to-end tests of Ludwig workflows
+    └── ludwig             - Unit tests. Subdirectories match ludwig/ structure
+```
+
+We are using `pytest` as our testing framework. For more information, see the [pytest docs](https://docs.pytest.org).
+
+!!! note
+
+    Ludwig's test coverage is a work in progress, and many modules do not have proper test coverage yet. Contributions
+    which get us closer to the goal of 100% test coverage will be welcomed!
 
 ## Checklist
 
-Before running tests, make sure
+Before running tests, make sure:
 
-1. Your environment is properly setup.
-2. You have write access on the machine. Some of the tests require saving data to disk.
+1. Your python environment is properly setup to run Ludwig.
+2. All required dependencies for testing are installed: `pip install ludwig[test]`
+3. You have write access on the machine. Some tests require saving temporary files to disk.
 
 ## Running tests
 
-To run all tests, just run
-`python -m pytest` from the ludwig root directory.
-Note that you don't need to have ludwig module installed and in this case
-code change will take effect immediately.
+To run all tests, execute `python -m pytest` from the ludwig root directory.
+Note that you don't need to have ludwig module installed. Running tests from the ludwig source root is useful for
+development as the test will import ludwig modules directly from the source tree.
 
-To run a single test, run
+To run all unit tests (will take a few minutes):
 
-```
-python -m pytest path_to_filename -k "test_method_name"
+```python
+python -m pytest tests/ludwig/
```

Comment:
No that was not the intent of that line, the intent was to show how to run a single specific text, not the whole suite. If I remember correctly -m was needed for loading as module and resolve dependencies and -k to specify the specific test. Please doublecheck

Reply:
Running `pytest` directly has been working great for me both locally and on Ray, for a directory of tests and specific tests. I'm not aware of any dependency issues. @dantreiman ?

Reply:
`pytest` works fine for me.

I wrote `python -m pytest` because it removes the potential point of failure where the pytest binary isn't in the path, or the user's path is misconfigured and some other version of pytest shadows it.

Again not a big deal to me.

Reply:
sounds good, as long as we show both hot to run all tests in a dir, a specific test module and a specific test within a module.